crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,648 @@
+    match src_addr.ip() {
+        IpAddr::V4(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+        IpAddr::V6(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+    };
+    src_addr.port().hash(&mut hasher);
+    hasher.finish() as u16 // TODO: think if we want to use the full 16 bits or maybe some range e.g. >=1024
+}
+
+pub struct StarcastNat {
+    mappings: LruCache<u16, SocketAddr>,
+    vpeer_ipv4_addr: Ipv4Addr,
+    vpeer_ipv6_addr: Ipv6Addr,
+    mapper: Mapper,
+}
+
+impl StarcastNat {
+    /// Create new starcast nat with default capcity of 2048 entries and 2 minute timeout
+    /// 2min default is taken from (https://datatracker.ietf.org/doc/html/rfc4787#section-4.3)
+    pub fn new(vpeer_ipv4: Ipv4Addr, vpeer_ipv6: Ipv6Addr) -> Self {
+        const LRU_CAPACITY: usize = 2048;
+        const LRU_TIMEOUT: Duration = Duration::from_secs(120);
+        Self::new_custom(
+            vpeer_ipv4,
+            vpeer_ipv6,
+            LRU_CAPACITY,
+            LRU_TIMEOUT,
+            default_mapper,
+        )
+    }
+
+    /// Create new starcast nat with custom capacity and timeout
+    fn new_custom(
+        vpeer_ipv4_addr: Ipv4Addr,
+        vpeer_ipv6_addr: Ipv6Addr,
+        capacity: usize,
+        timeout: Duration,
+        mapper: Mapper,
+    ) -> Self {
+        Self {
+            mappings: LruCache::new(timeout, capacity),
+            vpeer_ipv4_addr,
+            vpeer_ipv6_addr,
+            mapper,
+        }
+    }
+
+    fn translate_incomming_internal<'a, P: MutableIpPacket<'a>>(
+        &mut self,
+        packet: &'a mut [u8],
+    ) -> Result<(), Error> {
+        let mut ip_packet = P::new(packet).ok_or(Error::PacketTooShort)?;
+        if ip_packet.get_next_level_protocol() != IpNextHeaderProtocols::Udp {
+            return Err(Error::UnexpectedTransportProtocol);
+        }
+
+        let src_ip = ip_packet.get_source();
+        let dst_ip = ip_packet.get_destination();
+
+        match src_ip {
+            IpAddr::V4(_) => {
+                ip_packet.set_source(IpAddr::V4(self.vpeer_ipv4_addr));
+                ip_packet.set_checksum(ipv4::checksum(&ip_packet.to_immutable_ipv4()));
+            }
+            IpAddr::V6(_) => {
+                ip_packet.set_source(IpAddr::V6(self.vpeer_ipv6_addr));
+            }
+        };
+
+        let mut udp_packet =
+            MutableUdpPacket::new(ip_packet.payload_mut()).ok_or(Error::PacketTooShort)?;
+        let src_port = udp_packet.get_source();
+        let src_addr = SocketAddr::new(src_ip, src_port);
+        let mut nat_port = (self.mapper)(src_addr);
+
+        while let Some(&current_mapping) = self.mappings.peek(&nat_port) {
+            if current_mapping != src_addr {
+                nat_port = nat_port.wrapping_add(1);
+            } else {
+                break;
+            }
+        }
+
+        self.mappings.insert(nat_port, src_addr);
+
+        udp_packet.set_source(nat_port);
```

Comment:
upd packet also has it's own checksum, so it's would be better to move checksum ajusting after all of updates, and ajust them both at once

Reply:
Oh i think i missed that you do udp checksum under this :upside_down_face: